Parallel initialization of UTXO commitment

dagurval opened the pull request:
Use multiple threads to calculate UTXO commitment from a coins view. On my system it takes 65 seconds to calculate the utxo commitment for mainnet using `-par=16` (16 threads).

(This is on top of #439 and #425)

Suggest cherry-picking this into #425 @tomasvdw 
```
2018-06-03 14:12:30 Generating UTXO commitment; progress 0
2018-06-03 14:12:32 Generating UTXO commitment; progress 2
2018-06-03 14:12:33 Generating UTXO commitment; progress 4
2018-06-03 14:12:35 Generating UTXO commitment; progress 7
2018-06-03 14:12:36 Generating UTXO commitment; progress 9
2018-06-03 14:12:38 Generating UTXO commitment; progress 12
2018-06-03 14:12:40 Generating UTXO commitment; progress 14
2018-06-03 14:12:41 Generating UTXO commitment; progress 17
2018-06-03 14:12:43 Generating UTXO commitment; progress 19
2018-06-03 14:12:45 Generating UTXO commitment; progress 22
2018-06-03 14:12:46 Generating UTXO commitment; progress 24
2018-06-03 14:12:48 Generating UTXO commitment; progress 27
2018-06-03 14:12:50 Generating UTXO commitment; progress 29
2018-06-03 14:12:51 Generating UTXO commitment; progress 32
2018-06-03 14:12:53 Generating UTXO commitment; progress 34
2018-06-03 14:12:54 Generating UTXO commitment; progress 37
2018-06-03 14:12:56 Generating UTXO commitment; progress 39
2018-06-03 14:12:58 Generating UTXO commitment; progress 42
2018-06-03 14:12:59 Generating UTXO commitment; progress 44
2018-06-03 14:13:01 Generating UTXO commitment; progress 47
2018-06-03 14:13:02 Generating UTXO commitment; progress 49
2018-06-03 14:13:04 Generating UTXO commitment; progress 52
2018-06-03 14:13:06 Generating UTXO commitment; progress 54
2018-06-03 14:13:07 Generating UTXO commitment; progress 57
2018-06-03 14:13:09 Generating UTXO commitment; progress 59
2018-06-03 14:13:11 Generating UTXO commitment; progress 62
2018-06-03 14:13:12 Generating UTXO commitment; progress 64
2018-06-03 14:13:14 Generating UTXO commitment; progress 67
2018-06-03 14:13:16 Generating UTXO commitment; progress 69
2018-06-03 14:13:17 Generating UTXO commitment; progress 72
2018-06-03 14:13:19 Generating UTXO commitment; progress 75
2018-06-03 14:13:20 Generating UTXO commitment; progress 77
2018-06-03 14:13:22 Generating UTXO commitment; progress 80
2018-06-03 14:13:24 Generating UTXO commitment; progress 82
2018-06-03 14:13:25 Generating UTXO commitment; progress 85
2018-06-03 14:13:27 Generating UTXO commitment; progress 87
2018-06-03 14:13:29 Generating UTXO commitment; progress 90
2018-06-03 14:13:30 Generating UTXO commitment; progress 92
2018-06-03 14:13:32 Generating UTXO commitment; progress 94
2018-06-03 14:13:33 Generating UTXO commitment; progress 97
2018-06-03 14:13:35 utxo commit: 1a67734f41942d108ebd5149b69a5060f8c8a4261e1131be882745665f2b2b60
```

tomasvdw commented:
Nice! I will cherry pick it tomorrow.

dgenr8 commented:
I'd favor merging #425 as is, and this later.

dgenr8 commented:
This can be reduced to just the supporting cherry-picks and merged first.  Then #425 can be updated with the new parallelization, or that can be done on a new PR (#425 looks ready to me).

dagurval commented:
@dgenr8 sounds good.

I've rebased on top of #425 + some minor changes.

tomasvdw commented on src/utilutxocommit.cpp:
```diff
@@ -0,0 +1,124 @@
+    // spin up workers
+    auto worker = [&buffer, &eof]() {
+        CUtxoCommit commitment;
+        auto utxo = std::make_unique<std::pair<COutPoint, Coin>>();
```
make_unique seems to be the first C++14 feature in the codebase. Not sure if intentional, but maybe just stick to std::unqiue_ptr constructor?

tomasvdw commented on src/utilutxocommit.cpp:
```diff
@@ -0,0 +1,124 @@
+    // spin up workers
+    auto worker = [&buffer, &eof]() {
+        CUtxoCommit commitment;
+        auto utxo = std::make_unique<std::pair<COutPoint, Coin>>();
```
This seems to be the first use of a C++14 feature in the codebase. I am not sure if intentional, but maybe better stick to unique_ptr constructor?

dagurval replied:
😱

dagurval commented:
removed use of std::make_unique

tomasvdw commented:
Maybe it's a matter of taste, but I am not sure whether the custom circularbuffer has any benefit here; it's seems to be an unnecessary amount of plumbing. Why not use the cursor directly?

I've tried it with a simpler approach using local buffers and a mutex around cursor access and it seems both less plumbing and actually few percent faster. See: https://gist.github.com/tomasvdw/a9399c657f6dd2d0c5de30a57fa264c5

dagurval commented:
If your approach is equally fast or not significantly slower, then let's go for it. Simpler is better.

> `utxos.push_back(utxo);`

Looks like you're not staying within your 50 element reserved limit. Did you mean to either: 

`utxos[n] = utxo` + logic to not exceed n in the second loop at `break`

Or `utxos.clear()` at the end of while loop?

tomasvdw commented:
@dagurval The utxos.clear() at the end isn't needed as the utxos vector is local to the while loop. 

Maybe it's more efficient to use a single 50 elements array (declared outside the while) that gets overwritten each while iteration (taking n into account in the second loop for the break case)? Though that seems like a micro optimization which may not weigh against the readability...

dagurval commented:
> The utxos.clear() at the end isn't needed as the utxos vector is local to the while loop.

Oops, missed that :-)

tomasvdw commented:
Should I PR the alternative separately? Or do you want to make the changes here?

dagurval commented:
@tomasvdw I'll get around to it soon. Just finishing up something else. If you want to PR an alternative to speed tings up, that's fine with me.

dagurval commented:
I added some benchmarking code. This is the result when compiled in release.
_tomasvdw + circular buffer_ is where I tested @tomasvdw approach, but gave each worker a circular buffer instead of vector. This made little difference.

Surprisingly @tomasvdw is faster when compiled in debug, but slower with release.

```
# original
#Benchmark,count,min,max,average,min_cycles,max_cycles,average_cycles
BuildCommitment16,786432,0.000000241223461,0.000001531374437,0.000001389270741,866,5502,4991
BuildCommitment16,786432,0.000000618743798,0.000001532707756,0.000001395727850,2223,5507,5015
BuildCommitment8,655360,0.000000988598913,0.000001785963832,0.000001696049367,3552,6417,6094
BuildCommitment8,589824,0.000001083080861,0.000001951553713,0.000001723169892,3891,7012,6191
BuildCommitment4,327680,0.000003008313797,0.000003533554263,0.000003413064405,10809,12696,12263
BuildCommitment4,327680,0.000003016324627,0.000003600387572,0.000003435052349,10838,12937,12342

# tomasvdw
BuildCommitment16,720896,0.000001480191713,0.000001960812369,0.000001489155032,5318,7045,5350
BuildCommitment16,720896,0.000001481843356,0.000001887441613,0.000001492879660,5324,6781,5364
BuildCommitment8,589824,0.000001734832040,0.000002233748091,0.000001803804076,6233,8026,6481
BuildCommitment8,589824,0.000001733063982,0.000002512068022,0.000001790763134,6227,9026,6434
BuildCommitment4,294912,0.000003459252184,0.000003741210094,0.000003500586899,12429,13442,12578
BuildCommitment4,294912,0.000003462526365,0.000003738532541,0.000003503340445,12441,13433,12588

# tomasvdw + circular buffer
BuildCommitment16,720896,0.000001471693395,0.000001870241249,0.000001488487959,5288,6720,5348
BuildCommitment16,720896,0.000001477263140,0.000002001848770,0.000001485326886,5308,7193,5337
BuildCommitment8,589824,0.000001726544724,0.000002506480087,0.000001799404547,6203,9006,6465
BuildCommitment8,589824,0.000001726348273,0.000002474494977,0.000001822813323,6203,8891,6549
BuildCommitment4,294912,0.000003447297786,0.000003729248419,0.000003469760284,12386,13400,12467
BuildCommitment4,294912,0.000003436289262,0.000003746943548,0.000003467027758,12347,13463,12457
```
![bench](https://user-images.githubusercontent.com/92707/41913844-b962c76c-7951-11e8-86a3-b8ae109823c9.png)

dagurval commented:
I've added a benchmark tests and the benchmarks shows @tomasvdw method to be better. 

I changed the worker buffer from 50 to 500, as testing with `-buildcommitment=1` shows less lock contention. This is not visible on the benchmark though.

dagurval commented:
@tomasvdw does this PR look good to you now?
